File: training-data/php/newton_sqrt.php
<?php

function newtonSqrt(float $x, float $eps = 1e-12): float
{
    if ($x < 0) {
        throw new InvalidArgumentException('negative input');
    }
    if ($x === 0.0) {
        return 0.0;
    }
    $g = $x > 1 ? $x / 2 : 1.0;
    while (abs($g * $g - $x) > $eps * $x) {
        $g = ($g + $x / $g) / 2;
    }
    return $g;
}

function newtonRoot(callable $f, callable $df, float $x0, int $maxIter = 50): float
{
    for ($i = 0; $i < $maxIter; $i++) {
        $step = $f($x0) / $df($x0);
        $x0 -= $step;
        if (abs($step) < 1e-15) {
            break;
        }
    }
    return $x0;
}

foreach ([2, 9, 0.25, 1e10] as $x) {
    printf("sqrt(%g) = %.12f (builtin %.12f)\n", $x, newtonSqrt($x), sqrt($x));
}
printf("root of cos(x) - x: %.15f\n", newtonRoot(fn ($x) => cos($x) - $x, fn ($x) => -sin($x) - 1, 1.0));

File: training-data/perl/newton_sqrt.pl
use strict;
use warnings;

sub newton_sqrt {
    my ($x, $eps) = (@_, 1e-12);
    die "negative input\n" if $x < 0;
    return 0 if $x == 0;
    my $g = $x > 1 ? $x / 2 : 1;
    $g = ($g + $x / $g) / 2 while abs($g * $g - $x) > $eps * $x;
    return $g;
}

sub newton_root {
    my ($f, $df, $x) = @_;
    for (1 .. 50) {
        my $step = $f->($x) / $df->($x);
        $x -= $step;
        last if abs($step) < 1e-15;
    }
    return $x;
}

printf "sqrt(%g) = %.12f (builtin %.12f)\n", $_, newton_sqrt($_), sqrt $_ for 2, 9, 0.25, 1e10;
printf "root of cos(x) - x: %.15f\n", newton_root(sub { cos($_[0]) - $_[0] }, sub { -sin($_[0]) - 1 }, 1);
printf "cube root of 10: %.12f\n", newton_root(sub { $_[0]**3 - 10 }, sub { 3 * $_[0]**2 }, 2);
